### PR TITLE
WE-2532 - Add JWE Encryption Options to Ceramic Upload

### PR DESCRIPTION
**Pull request by Custard7**
[workerB](https://workerb.linearb.io/v2/badge/collaboration-page?magicLinkId=hsZsBZo)
# Overview

#### 📚 What is the context and goal of this PR?

Currently, all Ceramic uploads are public! But we need more options to allow encrypting content stored in Ceramic.

#### 🥴 TL; RL:

This PR enables JWE encryption of credentials by adding a new, optional control plane method: `uploadEncrypted`. This enables decryption of JWE credentials by specified DID recipients!

#### 💡 Feature Breakdown (screenshots & videos encouraged!)

See encrypt / decrypt of JWE credential:

Screen Shot 2022-10-31 at 12 11 00 PM

#### 🛠 Important tradeoffs made:

- A user can only specify the advanced Ceramic encryption params if they directly use the `learnCard.invoke.publishContentToCeramic(vc, encryption)`
- Metadata is not encrypted - this is currently a limitation of ceramic.
-  Anyone who implements uploadEncrypted in the plugin must abide by encrypting according to recipient DIDs.

#### 🔍 Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Chore (refactor, documentation update, etc)

#### 💳 Does This Create Any New Technical Debt? ( If yes, please describe and [add JIRA TODOs](https://welibrary.atlassian.net/jira/software/projects/WE/boards/2) )
- [ ] No
- [ ] Yes

# Testing

## Test 1 - upload/get encrypted credential.

```
const l = await initLearnCard({ seed: '123', debug: console.log })
const malicious = await initLearnCard({ seed: '666', debug: console.log });
const friend = await initLearnCard({ seed: '808', debug: console.log });

const uvc = l.invoke.getTestVc();   
const vc = await  l.invoke.issueCredential(uvc);
const streamId = await l.store.Ceramic.uploadEncrypted(vc);

// Should properly retrieve credential!  ✅
await l.read.get(streamId);

// Should error -  Could not decrypt credential - DID not authorized. ❌
await malicious.read.get(streamId);
```

## Test 2 - upload/get shared encrypted credential.
```
const sharedStreamId = await l.store.Ceramic.uploadEncrypted(vc, { recipients: ['did:key:z6Mki5xcJBP54S5zpUC4dK7ftREcMJwwRhWS9AQAf7p2McEb'] });

// Should properly retrieve credential!  ✅
await l.read.get(sharedStreamId);
// Should properly retrieve credential!  ✅
await friend.read.get(sharedStreamId);

// Should error -  Could not decrypt credential - DID not authorized. ❌
await malicious.read.get(sharedStreamId);
```

## Test 3 - upload/get one-way encrypted credential.
```
const unidirectionalStreamId = await l.invoke.publishContentToCeramic(vc, { encrypt: true, controllersCanDecrypt: false, recipients: ['did:key:z6Mki5xcJBP54S5zpUC4dK7ftREcMJwwRhWS9AQAf7p2McEb'] });

// Should properly retrieve credential!  ✅
await friend.read.get(unidirectionalStreamId);

// Should error -  Could not decrypt credential - DID not authorized. ❌
await l.read.get(unidirectionalStreamId);
// Should error -  Could not decrypt credential - DID not authorized. ❌
await malicious.read.get(unidirectionalStreamId);
```

#### 🧪 Code Coverage

# Documentation

#### 📜 Gitbook

https://app.gitbook.com/o/6uDv1QDlxaaZC7i8EaGb/s/FXvEJ9j3Vf3FW5Nc557n/~/changes/s5964VnR2VHP0jDIks2A/learn-card-sdk/learncard-core/plugins/official-plugins/ceramic

#### 📊 Storybook

# ✅ PR Checklist
- [x] Related to a Jira issue ([create one if not](https://welibrary.atlassian.net/jira/software/projects/WE/boards/2))
- [x] My code follows **style guidelines** (eslint / prettier)
- [x] I have **manually tested** common end-2-end cases
- [x] I have **reviewed** my code
- [x] I have **commented** my code, particularly where ambiguous
- [x] New and existing **unit tests pass** locally with my changes

Screen Shot 2022-10-31 at 12 46 52 PM

- [x] I have made corresponding changes to **gitbook documentation**

### 🚀 Ready to squash-and-merge?:
- [x] Code is backwards compatible
- [x] There is **not** a "Do Not Merge" label on this PR
- [x] I have thoughtfully considered the security implications of this change.
- [x] This change does not expose new public facing endpoints that do not have authentication

### CODE REVIEW

File: packages/learn-card-core/src/wallet/plugins/ceramic/ceramic.ts [hunk 1]
```diff
@@ -46,7 +49,8 @@ export const getCeramicPlugin = async <URI extends string = ''>(
     const publishContentToCeramic = async (
         content: any,
         metadata: TileMetadataArgs = {},
-        options: CreateOpts = {}
+        options: CreateOpts = {},
+        encryption: CeramicEncryptionParams | undefined
```
Comment (TaylorBeeston) [inline packages/learn-card-core/src/wallet/plugins/ceramic/ceramic.ts:53]:
```suggestion
        encryption?: CeramicEncryptionParams
```

nit: Clears up that ESLint error

File: packages/learn-card-core/src/wallet/plugins/ceramic/ceramic.ts [hunk 2]
```diff
@@ -114,17 +159,24 @@ export const getCeramicPlugin = async <URI extends string = ''>(
                 if (!verificationResult.success) return undefined;
 
                 const streamId = verificationResult.data.split(':')[2];
-
-                return VCValidator.parseAsync(await readContentFromCeramic(streamId));
+                try {
+                    return VCValidator.parseAsync(await readContentFromCeramic(streamId));
```
Comment (TaylorBeeston) [inline packages/learn-card-core/src/wallet/plugins/ceramic/ceramic.ts:163]:
```suggestion
                    return await VCValidator.parseAsync(await readContentFromCeramic(streamId));
```

nit: This is that weird edge case where we have to `await` before returning or else the `try` won't catch the error!

File: packages/learn-card-core/test/index.test.ts [hunk 3]
```diff
@@ -458,6 +458,71 @@ describe('LearnCard SDK', () => {
         });
     });
 
+    describe('Ceramic Plugin', () => {
```
[the remark this answers is not in the document]
Comment (TaylorBeeston) [reply packages/learn-card-core/test/index.test.ts:461]: Yup! I want to add maybe like a cron job or something that actually checks all the persistence tests every once in a while, but having them rum multiple times per push on every branch is going to be way too much false traffic on our Ceramic Node IMO and greatly slow down tests/deployment/everything, so I would go ahead and just skip these for now!

File: packages/learn-card-core/src/wallet/base/wallet.ts [hunk 4]
```diff
@@ -139,6 +139,19 @@ const addCachingToStorePlane = <
     },
     // TODO: Add caching to uploadMany
     ...('uploadMany' in plane ? { uploadMany: plane.uploadMany } : {}),
+    ...('uploadEncrypted' in plane
+        ? {
+              uploadEncrypted: async (_learnCard, vc, params, { cache = 'cache-first' } = {}) => {
+                  const uri = await plane.uploadEncrypted?.(_learnCard, vc, params);
```
[the remark this answers is not in the document]
Comment (TaylorBeeston) [reply packages/learn-card-core/src/wallet/base/wallet.ts:145]: Yeah, this is one of those annoying limitations of the TS compiler in its current form. You can clearly see that we are explicitly checking for `uploadEncrypted` in the plane here, but it still thinks it can't be certain because that call is within the scope of an arbitrary function, so technically, according to TS, it's possible that the plane object could be changed to remove that method between when this call happens and when the method is actually called.

It's annoying, but technically correct according to the types it is given. Another thing I've thought about doing here is to always implement the optional methods and then automatically fallback or error based on some passed in config or something, which is probably the right solution and would free us from this weird TS pedantry

**Review by TaylorBeeston — APPROVED** [review]
LGTM! Great job with the types and everything! You have peered into the underbelly of this library and come out the other side victorious! 😂
